File: src/gitscale/storage.py
"""Cloud storage backends for manifests (S3-compatible + GCS).

Uses raw httpx — no SDK dependencies. Auth is detected from the URL:
- *.s3.amazonaws.com, *.r2.cloudflarestorage.com, or anything else → AWS Sig V4
- storage.googleapis.com → GCS Bearer token

Object paths follow the layout:
  {base_url}/{host}/{owner}/{repo}/{revision}.json
"""

from __future__ import annotations

import datetime
import hashlib
import hmac
import json
import os
import re
from dataclasses import dataclass
from typing import TYPE_CHECKING, Any
from urllib.parse import urlparse

import httpx

from gitscale.urls import _extract_hostname, extract_owner_repo

if TYPE_CHECKING:
    from pathlib import Path


class StorageError(Exception):
    """Raised when a cloud storage operation fails."""


@dataclass(frozen=True, slots=True)
class HeadResult:
    """Result of a HEAD request."""

    exists: bool
    etag: str


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------


def object_url(
    storage_url: str, repo_url: str, revision: str
) -> str:
    """Build the full object URL for a given repo + revision."""
    hostname = _extract_hostname(repo_url)
    owner, repo = extract_owner_repo(repo_url)
    safe_rev = revision.replace("/", "_") if revision else "_default"
    base = storage_url.rstrip("/")
    return f"{base}/{hostname}/{owner}/{repo}/{safe_rev}.json"


def head_object(url: str) -> HeadResult:
    """HEAD an object. Returns existence and ETag."""
    if _is_gcs(url):
        return _gcs_head(url)
    return _s3_head(url)


def get_object(url: str) -> tuple[bytes, str] | None:
    """GET an object. Returns (body, etag) or None if 404."""
    if _is_gcs(url):
        return _gcs_get(url)
    return _s3_get(url)


def put_object(url: str, body: bytes) -> str:
    """PUT an object. Returns the ETag from the response."""
    if _is_gcs(url):
        return _gcs_put(url, body)
    return _s3_put(url, body)


def upload_metadata(
    storage_url: str,
    repo_url: str,
    revision: str,
    data: dict[str, Any],
) -> str:
    """Upload metadata JSON to cloud storage.

    Returns the full object URL.
    """
    url = object_url(storage_url, repo_url, revision)
    body = json.dumps(data, indent=2).encode()
    put_object(url, body)
    return url


def download_metadata(
    storage_url: str,
    repo_url: str,
    revision: str,
) -> dict[str, Any] | None:
    """Download metadata JSON from cloud storage.

    Returns None if the object doesn't exist (404).
    """
    url = object_url(storage_url, repo_url, revision)
    result = get_object(url)
    if result is None:
        return None
    body, _etag = result
    data: dict[str, Any] = json.loads(body)
    return data


def clone_manifest(
    storage_url: str,
    repo_url: str,
    revision: str,
    dest: Path,
) -> bool:
    """Download manifest and save to dest/manifest.json with ETags.

    Only downloads if the directory doesn't exist yet.
    Returns True if found and downloaded.
    """
    url = object_url(storage_url, repo_url, revision)
    result = get_object(url)
    if result is None:
        return False
    body, etag = result
    dest.mkdir(parents=True, exist_ok=True)
    (dest / "manifest.json").write_bytes(body)
    (dest / ".etag").write_text(etag, encoding="utf-8")
    (dest / ".etag-remote").write_text(etag, encoding="utf-8")
    _apply_manifest_readonly(dest)
    return True


def fetch_manifest(
    storage_url: str,
    repo_url: str,
    revision: str,
    dest: Path,
) -> HeadResult:
    """HEAD the remote manifest and save .etag-remote."""
    url = object_url(storage_url, repo_url, revision)
    result = head_object(url)
    if result.exists:
        dest.mkdir(parents=True, exist_ok=True)
        (dest / ".etag-remote").write_text(result.etag, encoding="utf-8")
    return result


def pull_manifest(
    storage_url: str,
    repo_url: str,
    revision: str,
    dest: Path,
) -> bool:
    """HEAD + conditional GET for a manifest.

    Downloads only if remote ETag differs from local.
    Returns True if the manifest exists remotely.
    """
    url = object_url(storage_url, repo_url, revision)
    hr = head_object(url)
    if not hr.exists:
        return False

    dest.mkdir(parents=True, exist_ok=True)
    (dest / ".etag-remote").write_text(hr.etag, encoding="utf-8")

    # Check if local is already up to date
    local_etag_file = dest / ".etag"
    if local_etag_file.is_file():
        local_etag = local_etag_file.read_text(encoding="utf-8").strip()
        if local_etag == hr.etag:
            return True  # up to date

    # Download
    result = get_object(url)
    if result is None:
        return False
    body, etag = result
    _restore_manifest_writable(dest)
    (dest / "manifest.json").write_bytes(body)
    (dest / ".etag").write_text(etag, encoding="utf-8")
    _apply_manifest_readonly(dest)
    return True


def push_manifest(
    storage_url: str,
    repo_url: str,
    revision: str,
    dest: Path,
) -> bool:
    """HEAD + conditional PUT for a manifest.

    Uploads only if remote ETag differs from local.
    Returns True if upload happened, False if already up to date.
    """
    manifest_file = dest / "manifest.json"
    if not manifest_file.is_file():
        return False

    url = object_url(storage_url, repo_url, revision)

    # Check remote
    hr = head_object(url)
    local_etag_file = dest / ".etag"
    local_etag = ""
    if local_etag_file.is_file():
        local_etag = local_etag_file.read_text(encoding="utf-8").strip()

    if hr.exists and hr.etag == local_etag:
        # Remote matches local — nothing to push
        (dest / ".etag-remote").write_text(hr.etag, encoding="utf-8")
        return False

    body = manifest_file.read_bytes()
    etag = put_object(url, body)
    (dest / ".etag").write_text(etag, encoding="utf-8")
    (dest / ".etag-remote").write_text(etag, encoding="utf-8")
    return True


def _apply_manifest_readonly(dest: Path) -> None:
    """Remove write permission from manifest files."""
    import stat

    for name in ("manifest.json",):
        fpath = dest / name
        if fpath.is_file():
            mode = fpath.stat().st_mode
            fpath.chmod(mode & ~(stat.S_IWUSR | stat.S_IWGRP | stat.S_IWOTH))


def _restore_manifest_writable(dest: Path) -> None:
    """Restore write permission on manifest files before update."""
    import stat

    for name in ("manifest.json",):
        fpath = dest / name
        if fpath.is_file():
            mode = fpath.stat().st_mode
            fpath.chmod(mode | stat.S_IWUSR)


def _object_url(
    storage_url: str, repo_url: str, revision: str
) -> str:
    """Deprecated alias for object_url."""
    return object_url(storage_url, repo_url, revision)


# ---------------------------------------------------------------------------
# Backend detection
# ---------------------------------------------------------------------------


def _is_gcs(url: str) -> bool:
    """Check if the URL points to Google Cloud Storage."""
    parsed = urlparse(url)
    host = parsed.hostname or ""
    return host == "storage.googleapis.com" or host.endswith(
        ".storage.googleapis.com"
    )


# ---------------------------------------------------------------------------
# GET / PUT / HEAD dispatch — now handled by public API directly
# ---------------------------------------------------------------------------


# ---------------------------------------------------------------------------
# GCS (Bearer token)
# ---------------------------------------------------------------------------


def _gcs_token() -> str:
    """Resolve GCS bearer token from environment."""
    token = os.environ.get("GOOGLE_TOKEN") or os.environ.get(
        "GCLOUD_ACCESS_TOKEN"
    )
    if token:
        return token
    raise StorageError(
        "No GCS token found. Set GOOGLE_TOKEN or GCLOUD_ACCESS_TOKEN."
    )


def _gcs_head(url: str) -> HeadResult:
    token = _gcs_token()
    resp = httpx.head(
        url,
        headers={"Authorization": f"Bearer {token}"},
        timeout=30,
    )
    if resp.status_code == 404:
        return HeadResult(exists=False, etag="")
    _check_storage_response(resp, "HEAD")
    return HeadResult(
        exists=True, etag=resp.headers.get("etag", "")
    )


def _gcs_get(url: str) -> tuple[bytes, str] | None:
    token = _gcs_token()
    resp = httpx.get(
        url,
        headers={"Authorization": f"Bearer {token}"},
        timeout=30,
    )
    if resp.status_code == 404:
        return None
    _check_storage_response(resp, "GET")
    return resp.content, resp.headers.get("etag", "")


def _gcs_put(url: str, body: bytes) -> str:
    token = _gcs_token()
    resp = httpx.put(
        url,
        content=body,
        headers={
            "Authorization": f"Bearer {token}",
            "Content-Type": "application/json",
        },
        timeout=30,
    )
    _check_storage_response(resp, "PUT")
    return str(resp.headers.get("etag", ""))


# ---------------------------------------------------------------------------
# S3-compatible (AWS Signature V4)
# ---------------------------------------------------------------------------


def _s3_credentials() -> tuple[str, str]:
    """Resolve S3 access key and secret from environment."""
    access_key = os.environ.get("AWS_ACCESS_KEY_ID", "")
    secret_key = os.environ.get("AWS_SECRET_ACCESS_KEY", "")
    if not access_key or not secret_key:
        raise StorageError(
            "No S3 credentials found. "
            "Set AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY."
        )
    return access_key, secret_key


def _s3_region(url: str) -> str:
    """Extract region from URL or fall back to env/default."""
    env_region = os.environ.get("AWS_DEFAULT_REGION", "")
    if env_region:
        return env_region

    # Try to extract from s3.{region}.amazonaws.com
    parsed = urlparse(url)
    host = parsed.hostname or ""
    m = re.search(r"s3[.-]([a-z0-9-]+)\.amazonaws\.com", host)
    if m:
        region = m.group(1)
        if region != "amazonaws":
            return region

    return "us-east-1"


def _s3_sign(
    method: str,
    url: str,
    body: bytes,
    access_key: str,
    secret_key: str,
    region: str,
) -> dict[str, str]:
    """Sign an S3 request using AWS Signature V4.

    Returns the headers to add to the request.
    """
    parsed = urlparse(url)
    host = parsed.hostname or ""
    path = parsed.path or "/"

    now = datetime.datetime.now(datetime.UTC)
    datestamp = now.strftime("%Y%m%d")
    amzdate = now.strftime("%Y%m%dT%H%M%SZ")

    payload_hash = hashlib.sha256(body).hexdigest()

    # Canonical headers (must be sorted)
    canonical_headers = (
        f"host:{host}\n"
        f"x-amz-content-sha256:{payload_hash}\n"
        f"x-amz-date:{amzdate}\n"
    )
    signed_headers = "host;x-amz-content-sha256;x-amz-date"

    # Canonical request
    canonical_request = (
        f"{method}\n"
        f"{path}\n"
        f"\n"  # empty query string
        f"{canonical_headers}\n"
        f"{signed_headers}\n"
        f"{payload_hash}"
    )

    # String to sign
    scope = f"{datestamp}/{region}/s3/aws4_request"
    string_to_sign = (
        f"AWS4-HMAC-SHA256\n"
        f"{amzdate}\n"
        f"{scope}\n"
        f"{hashlib.sha256(canonical_request.encode()).hexdigest()}"
    )

    # Signing key
    def _sign(key: bytes, msg: str) -> bytes:
        return hmac.new(key, msg.encode(), hashlib.sha256).digest()

    k_date = _sign(f"AWS4{secret_key}".encode(), datestamp)
    k_region = _sign(k_date, region)
    k_service = _sign(k_region, "s3")
    k_signing = _sign(k_service, "aws4_request")

    signature = hmac.new(
        k_signing, string_to_sign.encode(), hashlib.sha256
    ).hexdigest()

    authorization = (
        f"AWS4-HMAC-SHA256 "
        f"Credential={access_key}/{scope}, "
        f"SignedHeaders={signed_headers}, "
        f"Signature={signature}"
    )

    return {
        "Authorization": authorization,
        "x-amz-date": amzdate,
        "x-amz-content-sha256": payload_hash,
        "Host": host,
    }


def _s3_head(url: str) -> HeadResult:
    access_key, secret_key = _s3_credentials()
    region = _s3_region(url)
    headers = _s3_sign("HEAD", url, b"", access_key, secret_key, region)

    resp = httpx.head(url, headers=headers, timeout=30)
    if resp.status_code == 404:
        return HeadResult(exists=False, etag="")
    _check_storage_response(resp, "HEAD")
    return HeadResult(
        exists=True, etag=resp.headers.get("etag", "")
    )


def _s3_get(url: str) -> tuple[bytes, str] | None:
    access_key, secret_key = _s3_credentials()
    region = _s3_region(url)
    headers = _s3_sign("GET", url, b"", access_key, secret_key, region)

    resp = httpx.get(url, headers=headers, timeout=30)
    if resp.status_code == 404:
        return None
    _check_storage_response(resp, "GET")
    return resp.content, resp.headers.get("etag", "")


def _s3_put(url: str, body: bytes) -> str:
    access_key, secret_key = _s3_credentials()
    region = _s3_region(url)
    headers = _s3_sign("PUT", url, body, access_key, secret_key, region)
    headers["Content-Type"] = "application/json"

    resp = httpx.put(url, content=body, headers=headers, timeout=30)
    _check_storage_response(resp, "PUT")
    return str(resp.headers.get("etag", ""))


# ---------------------------------------------------------------------------
# Shared
# ---------------------------------------------------------------------------


def _check_storage_response(resp: httpx.Response, method: str) -> None:
    """Raise StorageError on non-success responses."""
    if resp.status_code in (200, 201, 204):
        return
    raise StorageError(
        f"Storage {method} failed ({resp.status_code}): "
        f"{resp.text[:300]}"
    )

File: src/storage.rs
use anyhow::{bail, Result};
use chrono::Utc;
use hmac::{Hmac, Mac};
use regex::Regex;
use sha2::{Digest, Sha256};
use std::fs;
use std::os::unix::fs::PermissionsExt;
use std::path::{Path, PathBuf};
use url::Url;

use crate::urls::{extract_hostname, extract_owner_repo};

#[derive(Debug, Clone)]
pub struct HeadResult {
    pub exists: bool,
    pub etag: String,
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

pub fn object_url(storage_url: &str, repo_url: &str, revision: &str) -> Result<String> {
    let hostname = extract_hostname(repo_url)?;
    let (owner, repo) = extract_owner_repo(repo_url)?;
    let safe_rev = if revision.is_empty() {
        "_default".to_string()
    } else {
        revision.replace('/', "_")
    };
    let base = storage_url.trim_end_matches('/');
    Ok(format!(
        "{}/{}/{}/{}/{}.tar.gz",
        base, hostname, owner, repo, safe_rev
    ))
}

pub fn head_object(url: &str) -> Result<HeadResult> {
    if is_local(url) {
        local_head(url)
    } else if is_gcs(url) {
        gcs_head(url)
    } else {
        s3_head(url)
    }
}

pub fn get_object(url: &str) -> Result<Option<(Vec<u8>, String)>> {
    if is_local(url) {
        local_get(url)
    } else if is_gcs(url) {
        gcs_get(url)
    } else {
        s3_get(url)
    }
}

pub fn put_object(url: &str, body: &[u8]) -> Result<String> {
    if is_local(url) {
        local_put(url, body)
    } else if is_gcs(url) {
        gcs_put(url, body)
    } else {
        s3_put(url, body)
    }
}

// ---------------------------------------------------------------------------
// Artefact operations
// ---------------------------------------------------------------------------

pub fn clone_artefact(
    storage_url: &str,
    repo_url: &str,
    revision: &str,
    dest: &Path,
) -> Result<bool> {
    let url = object_url(storage_url, repo_url, revision)?;
    let Some((body, etag)) = get_object(&url)? else {
        return Ok(false);
    };
    fs::create_dir_all(dest)?;
    fs::write(dest.join(".etag"), &etag)?;
    fs::write(dest.join(".etag-remote"), &etag)?;
    extract_artefact(&body, dest)?;
    apply_artefact_readonly(dest)?;
    Ok(true)
}

pub fn fetch_artefact(
    storage_url: &str,
    repo_url: &str,
    revision: &str,
    dest: &Path,
) -> Result<HeadResult> {
    let url = object_url(storage_url, repo_url, revision)?;
    let result = head_object(&url)?;
    if result.exists {
        fs::create_dir_all(dest)?;
        fs::write(dest.join(".etag-remote"), &result.etag)?;
    }
    Ok(result)
}

pub fn pull_artefact(
    storage_url: &str,
    repo_url: &str,
    revision: &str,
    dest: &Path,
) -> Result<bool> {
    let url = object_url(storage_url, repo_url, revision)?;
    let hr = head_object(&url)?;
    if !hr.exists {
        return Ok(false);
    }
    fs::create_dir_all(dest)?;
    fs::write(dest.join(".etag-remote"), &hr.etag)?;

    // Check if local is already up to date
    let local_etag_file = dest.join(".etag");
    if local_etag_file.is_file() {
        let local_etag = fs::read_to_string(&local_etag_file)?.trim().to_string();
        if local_etag == hr.etag {
            return Ok(true); // up to date
        }
    }

    let Some((body, etag)) = get_object(&url)? else {
        return Ok(false);
    };
    restore_artefact_writable(dest)?;
    clean_artefact_files(dest)?;
    extract_artefact(&body, dest)?;
    fs::write(dest.join(".etag"), &etag)?;
    apply_artefact_readonly(dest)?;
    Ok(true)
}

// ---------------------------------------------------------------------------
// Artefact helpers
// ---------------------------------------------------------------------------

fn extract_artefact(body: &[u8], dest: &Path) -> Result<()> {
    use flate2::read::GzDecoder;
    use tar::Archive;

    let decoder = GzDecoder::new(body);
    let mut archive = Archive::new(decoder);
    let dest_canonical = dest.canonicalize()?;

    for entry in archive.entries()? {
        let mut entry = entry?;
        let entry_path = entry.path()?.to_path_buf();
        let resolved = dest
            .join(&entry_path)
            .canonicalize()
            .unwrap_or_else(|_| dest.join(&entry_path));

        // Security: reject paths that escape dest
        if !resolved.starts_with(&dest_canonical) {
            bail!(
                "artefact archive contains unsafe path: {}",
                entry_path.display()
            );
        }
        entry.unpack_in(dest)?;
    }
    Ok(())
}

fn apply_artefact_readonly(dest: &Path) -> Result<()> {
    for entry in fs::read_dir(dest)? {
        let entry = entry?;
        let path = entry.path();
        if path.is_file() && !entry.file_name().to_string_lossy().starts_with('.') {
            let meta = fs::metadata(&path)?;
            let mut perms = meta.permissions();
            perms.set_mode(perms.mode() & !(0o222));
            fs::set_permissions(&path, perms)?;
        }
    }
    Ok(())
}

fn restore_artefact_writable(dest: &Path) -> Result<()> {
    for entry in fs::read_dir(dest)? {
        let entry = entry?;
        let path = entry.path();
        if path.is_file() && !entry.file_name().to_string_lossy().starts_with('.') {
            let meta = fs::metadata(&path)?;
            let mut perms = meta.permissions();
            perms.set_mode(perms.mode() | 0o200);
            fs::set_permissions(&path, perms)?;
        }
    }
    Ok(())
}

fn clean_artefact_files(dest: &Path) -> Result<()> {
    for entry in fs::read_dir(dest)? {
        let entry = entry?;
        let name = entry.file_name();
        if name.to_string_lossy().starts_with('.') {
            continue;
        }
        let path = entry.path();
        if path.is_dir() {
            fs::remove_dir_all(&path)?;
        } else {
            fs::remove_file(&path)?;
        }
    }
    Ok(())
}

// ---------------------------------------------------------------------------
// Backend detection
// ---------------------------------------------------------------------------

fn is_local(url: &str) -> bool {
    url.starts_with("file://") || url.starts_with('/') || url.starts_with("./")
}

fn is_gcs(url: &str) -> bool {
    if let Ok(parsed) = Url::parse(url) {
        if let Some(host) = parsed.host_str() {
            return host == "storage.googleapis.com" || host.ends_with(".storage.googleapis.com");
        }
    }
    false
}

// ---------------------------------------------------------------------------
// Local filesystem
// ---------------------------------------------------------------------------

fn local_fs_path(url: &str) -> PathBuf {
    PathBuf::from(url.strip_prefix("file://").unwrap_or(url))
}

fn local_content_etag(content: &[u8]) -> String {
    format!("\"{}\"", hex::encode(Sha256::digest(content)))
}

fn local_head(url: &str) -> Result<HeadResult> {
    let path = local_fs_path(url);
    if !path.is_file() {
        return Ok(HeadResult {
            exists: false,
            etag: String::new(),
        });
    }
    let content = fs::read(&path)?;
    let etag = local_content_etag(&content);
    Ok(HeadResult { exists: true, etag })
}

fn local_get(url: &str) -> Result<Option<(Vec<u8>, String)>> {
    let path = local_fs_path(url);
    if !path.is_file() {
        return Ok(None);
    }
    let content = fs::read(&path)?;
    let etag = local_content_etag(&content);
    Ok(Some((content, etag)))
}

fn local_put(url: &str, body: &[u8]) -> Result<String> {
    let path = local_fs_path(url);
    if let Some(parent) = path.parent() {
        fs::create_dir_all(parent)?;
    }
    fs::write(&path, body)?;
    Ok(local_content_etag(body))
}

// ---------------------------------------------------------------------------
// GCS (Bearer token)
// ---------------------------------------------------------------------------

fn gcs_token() -> Result<String> {
    if let Ok(token) = std::env::var("GOOGLE_TOKEN") {
        if !token.is_empty() {
            return Ok(token);
        }
    }
    if let Ok(token) = std::env::var("GCLOUD_ACCESS_TOKEN") {
        if !token.is_empty() {
            return Ok(token);
        }
    }
    bail!("No GCS token found. Set GOOGLE_TOKEN or GCLOUD_ACCESS_TOKEN.")
}

fn gcs_head(url: &str) -> Result<HeadResult> {
    let token = gcs_token()?;
    let client = reqwest::blocking::Client::new();
    let resp = client
        .head(url)
        .header("Authorization", format!("Bearer {}", token))
        .timeout(std::time::Duration::from_secs(30))
        .send()?;
    if resp.status().as_u16() == 404 {
        return Ok(HeadResult {
            exists: false,
            etag: String::new(),
        });
    }
    check_storage_response(&resp, "HEAD")?;
    let etag = resp
        .headers()
        .get("etag")
        .and_then(|v| v.to_str().ok())
        .unwrap_or("")
        .to_string();
    Ok(HeadResult { exists: true, etag })
}

fn gcs_get(url: &str) -> Result<Option<(Vec<u8>, String)>> {
    let token = gcs_token()?;
    let client = reqwest::blocking::Client::new();
    let resp = client
        .get(url)
        .header("Authorization", format!("Bearer {}", token))
        .timeout(std::time::Duration::from_secs(30))
        .send()?;
    if resp.status().as_u16() == 404 {
        return Ok(None);
    }
    check_storage_response(&resp, "GET")?;
    let etag = resp
        .headers()
        .get("etag")
        .and_then(|v| v.to_str().ok())
        .unwrap_or("")
        .to_string();
    let body = resp.bytes()?.to_vec();
    Ok(Some((body, etag)))
}

fn gcs_put(url: &str, body: &[u8]) -> Result<String> {
    let token = gcs_token()?;
    let client = reqwest::blocking::Client::new();
    let resp = client
        .put(url)
        .header("Authorization", format!("Bearer {}", token))
        .header("Content-Type", "application/json")
        .body(body.to_vec())
        .timeout(std::time::Duration::from_secs(30))
        .send()?;
    check_storage_response(&resp, "PUT")?;
    let etag = resp
        .headers()
        .get("etag")
        .and_then(|v| v.to_str().ok())
        .unwrap_or("")
        .to_string();
    Ok(etag)
}

// ---------------------------------------------------------------------------
// S3 (AWS Signature V4)
// ---------------------------------------------------------------------------

fn s3_credentials() -> Result<(String, String)> {
    let access_key = std::env::var("AWS_ACCESS_KEY_ID").unwrap_or_default();
    let secret_key = std::env::var("AWS_SECRET_ACCESS_KEY").unwrap_or_default();
    if access_key.is_empty() || secret_key.is_empty() {
        bail!("No S3 credentials found. Set AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY.");
    }
    Ok((access_key, secret_key))
}

fn s3_region(url: &str) -> String {
    if let Ok(region) = std::env::var("AWS_DEFAULT_REGION") {
        if !region.is_empty() {
            return region;
        }
    }
    if let Ok(parsed) = Url::parse(url) {
        if let Some(host) = parsed.host_str() {
            let re = Regex::new(r"s3[.\-]([a-z0-9-]+)\.amazonaws\.com").unwrap();
            if let Some(caps) = re.captures(host) {
                let region = &caps[1];
                if region != "amazonaws" {
                    return region.to_string();
                }
            }
        }
    }
    "us-east-1".to_string()
}

type HmacSha256 = Hmac<Sha256>;

fn hmac_sign(key: &[u8], msg: &str) -> Vec<u8> {
    let mut mac = HmacSha256::new_from_slice(key).expect("HMAC key length");
    mac.update(msg.as_bytes());
    mac.finalize().into_bytes().to_vec()
}

fn s3_sign(
    method: &str,
    url: &str,
    body: &[u8],
    access_key: &str,
    secret_key: &str,
    region: &str,
) -> Vec<(String, String)> {
    let parsed = Url::parse(url).expect("valid URL");
    let host = parsed.host_str().unwrap_or("");
    let path = parsed.path();

    let now = Utc::now();
    let datestamp = now.format("%Y%m%d").to_string();
    let amzdate = now.format("%Y%m%dT%H%M%SZ").to_string();

    let payload_hash = hex::encode(Sha256::digest(body));

    // Canonical headers (sorted)
    let canonical_headers = format!(
        "host:{}\nx-amz-content-sha256:{}\nx-amz-date:{}\n",
        host, payload_hash, amzdate
    );
    let signed_headers = "host;x-amz-content-sha256;x-amz-date";

    // Canonical request
    let canonical_request = format!(
        "{}\n{}\n\n{}\n{}\n{}",
        method, path, canonical_headers, signed_headers, payload_hash
    );

    // String to sign
    let scope = format!("{}/{}/s3/aws4_request", datestamp, region);
    let canonical_hash = hex::encode(Sha256::digest(canonical_request.as_bytes()));
    let string_to_sign = format!(
        "AWS4-HMAC-SHA256\n{}\n{}\n{}",
        amzdate, scope, canonical_hash
    );

    // Signing key
    let k_date = hmac_sign(format!("AWS4{}", secret_key).as_bytes(), &datestamp);
    let k_region = hmac_sign(&k_date, region);
    let k_service = hmac_sign(&k_region, "s3");
    let k_signing = hmac_sign(&k_service, "aws4_request");

    let signature = hex::encode(hmac_sign(&k_signing, &string_to_sign));

    let authorization = format!(
        "AWS4-HMAC-SHA256 Credential={}/{}, SignedHeaders={}, Signature={}",
        access_key, scope, signed_headers, signature
    );

    vec![
        ("Authorization".to_string(), authorization),
        ("x-amz-date".to_string(), amzdate),
        ("x-amz-content-sha256".to_string(), payload_hash),
        ("Host".to_string(), host.to_string()),
    ]
}

fn s3_head(url: &str) -> Result<HeadResult> {
    let (access_key, secret_key) = s3_credentials()?;
    let region = s3_region(url);
    let headers = s3_sign("HEAD", url, b"", &access_key, &secret_key, &region);

    let client = reqwest::blocking::Client::new();
    let mut req = client.head(url).timeout(std::time::Duration::from_secs(30));
    for (k, v) in &headers {
        req = req.header(k.as_str(), v.as_str());
    }
    let resp = req.send()?;
    if resp.status().as_u16() == 404 {
        return Ok(HeadResult {
            exists: false,
            etag: String::new(),
        });
    }
    check_storage_response(&resp, "HEAD")?;
    let etag = resp
        .headers()
        .get("etag")
        .and_then(|v| v.to_str().ok())
        .unwrap_or("")
        .to_string();
    Ok(HeadResult { exists: true, etag })
}

fn s3_get(url: &str) -> Result<Option<(Vec<u8>, String)>> {
    let (access_key, secret_key) = s3_credentials()?;
    let region = s3_region(url);
    let headers = s3_sign("GET", url, b"", &access_key, &secret_key, &region);

    let client = reqwest::blocking::Client::new();
    let mut req = client.get(url).timeout(std::time::Duration::from_secs(30));
    for (k, v) in &headers {
        req = req.header(k.as_str(), v.as_str());
    }
    let resp = req.send()?;
    if resp.status().as_u16() == 404 {
        return Ok(None);
    }
    check_storage_response(&resp, "GET")?;
    let etag = resp
        .headers()
        .get("etag")
        .and_then(|v| v.to_str().ok())
        .unwrap_or("")
        .to_string();
    let body = resp.bytes()?.to_vec();
    Ok(Some((body, etag)))
}

fn s3_put(url: &str, body: &[u8]) -> Result<String> {
    let (access_key, secret_key) = s3_credentials()?;
    let region = s3_region(url);
    let headers = s3_sign("PUT", url, body, &access_key, &secret_key, &region);

    let client = reqwest::blocking::Client::new();
    let mut req = client
        .put(url)
        .body(body.to_vec())
        .header("Content-Type", "application/json")
        .timeout(std::time::Duration::from_secs(30));
    for (k, v) in &headers {
        req = req.header(k.as_str(), v.as_str());
    }
    let resp = req.send()?;
    check_storage_response(&resp, "PUT")?;
    let etag = resp
        .headers()
        .get("etag")
        .and_then(|v| v.to_str().ok())
        .unwrap_or("")
        .to_string();
    Ok(etag)
}

// ---------------------------------------------------------------------------
// Shared
// ---------------------------------------------------------------------------

fn check_storage_response(resp: &reqwest::blocking::Response, method: &str) -> Result<()> {
    let status = resp.status().as_u16();
    if matches!(status, 200 | 201 | 204) {
        return Ok(());
    }
    bail!(
        "Storage {} failed ({}): response status {}",
        method,
        status,
        status
    );
}
